test/Symmetric/OpensslAeadTest.php

```diff
@@ -0,0 +1,211 @@
+
+        $plaintext = Rand::getBytes(1024);
+        $encrypt = $this->crypt->encrypt($plaintext);
+        $tag = $this->crypt->getTag();
+
+        $this->assertEquals($this->crypt->getTagSize(), strlen($tag));
+        $this->assertEquals(mb_substr($encrypt, 0, $this->crypt->getTagSize(), '8bit'), $tag);
+
+        $decrypt = $this->crypt->decrypt($encrypt);
+        $tag2 = $this->crypt->getTag();
+        $this->assertEquals($tag, $tag2);
+        $this->assertEquals($plaintext, $decrypt);
+    }
+
+    /**
+     * @dataProvider getAuthEncryptionMode
+     * @expectedException Zend\Crypt\Symmetric\Exception\RuntimeException
+     */
+    public function testAuthenticationError($mode)
+    {
+        $this->crypt->setMode($mode);
+        $this->crypt->setKey(random_bytes($this->crypt->getKeySize()));
+        $this->crypt->setSalt(random_bytes($this->crypt->getSaltSize()));
+
+        $plaintext = Rand::getBytes(1024);
+        $encrypt = $this->crypt->encrypt($plaintext);
+
+        // Alter the encrypted message
+        $i = rand(0, mb_strlen($encrypt, '8bit') - 1);
+        $encrypt[$i] = $encrypt[$i] ^ chr(1);
+
+        $decrypt = $this->crypt->decrypt($encrypt);
+    }
+
+    public function testGcmEncryptWithTagSize()
+    {
+        $this->crypt->setMode('gcm');
+        $this->crypt->setKey(random_bytes($this->crypt->getKeySize()));
+        $this->crypt->setSalt(random_bytes($this->crypt->getSaltSize()));
+        $this->crypt->setTagSize(14);
+
+        $plaintext = Rand::getBytes(1024);
+        $encrypt = $this->crypt->encrypt($plaintext);
+        $this->assertEquals(14, $this->crypt->getTagSize());
+        $this->assertEquals($this->crypt->getTagSize(), strlen($this->crypt->getTag()));
+    }
+
+    public function testCcmEncryptWithTagSize()
+    {
+        $this->crypt->setMode('ccm');
+        $this->crypt->setKey(random_bytes($this->crypt->getKeySize()));
+        $this->crypt->setSalt(random_bytes($this->crypt->getSaltSize()));
+        $this->crypt->setTagSize(24);
+
+        $plaintext = Rand::getBytes(1024);
+        $encrypt = $this->crypt->encrypt($plaintext);
+        $this->assertEquals(24, $this->crypt->getTagSize());
+        $this->assertEquals($this->crypt->getTagSize(), strlen($this->crypt->getTag()));
+    }
+
+    /**
+     * @dataProvider getAuthEncryptionMode
+     */
+    public function testAuthenticatedEncryptionWithAdditionalData($mode)
+    {
+        $this->crypt->setMode($mode);
+        $this->crypt->setKey(random_bytes($this->crypt->getKeySize()));
+        $this->crypt->setSalt(random_bytes($this->crypt->getSaltSize()));
+        $this->crypt->setAad('foo@bar.com');
+
+        $plaintext = Rand::getBytes(1024);
+        $encrypt = $this->crypt->encrypt($plaintext);
+        $tag = $this->crypt->getTag();
+
+        $this->assertEquals($this->crypt->getTagSize(), strlen($tag));
+        $this->assertEquals(mb_substr($encrypt, 0, $this->crypt->getTagSize(), '8bit'), $tag);
+
```

Comment:
I changed strlen() usage in favor of mb_strlen(). Thanks for your library, I'll use it in the future!